### PR TITLE
Versal bringup

### PR DESCRIPTION
**Pull request by sjaeckel**
# Description

Complete port to Xilinx Versal with support for the AES-GCM, ECDSA, RSA, TRNG and SHA3-384 engines.

~The first part of this PR is handled in #5165~ **Edit**: This is now all handled in this PR.

# Testing

Tested on the VMK180.

# Checklist

 - [x] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: wolfcrypt/src/ecc.c [hunk 1]
```diff
@@ -7804,6 +8004,27 @@ int wc_ecc_verify_hash_ex(mp_int *r, mp_int *s, const byte* hash,
...
+    WOLFSSL_XIL_DCACHE_FLUSH_RANGE(XIL_CAST_U64(hashcopy), keySz);
+    WOLFSSL_XIL_DCACHE_FLUSH_RANGE(XIL_CAST_U64(key->key_raw), keySz * 2);
+    WOLFSSL_XIL_DCACHE_FLUSH_RANGE(XIL_CAST_U64(sigRS), keySz * 2);
+
+    err = XSecure_EllipticVerifySign(&(key->xSec.cinst), xil_curve_type[key->dp->id],
+                                     XIL_CAST_U64(hashcopy), keySz,
+                                     XIL_CAST_U64(key->key_raw),
+                                     XIL_CAST_U64(sigRS));
+
+    if (err) {
```
Comment (JacobBarthelmeh) [inline wolfcrypt/src/ecc.c:8022]: Lets check for the specific Xilinx return value (XST_SUCCESS ) to see if it was successful. Here and other places where XSecure_* API is called with Versal, makes the code more future proof and could help with if adding additional debug information on fail cases.

File: wolfcrypt/src/ecc.c [hunk 2]
```diff
@@ -10275,6 +10506,27 @@ static int wc_ecc_import_raw_private(ecc_key* key, const char* qx,
...
         }
     }
+#elif defined(WOLFSSL_XILINX_CRYPT_VERSAL)
+    if (err == MP_OKAY) {
+        const word32 key_size = key->dp->size;
+        word32 buf_size = sizeof(key->key_raw);
+        err = wc_export_int(key->pubkey.x, key->key_raw, &buf_size, key_size, WC_TYPE_UNSIGNED_BIN);
+        if (err == MP_OKAY) {
+            const word32 offset = buf_size;
+            buf_size = sizeof(key->key_raw) - offset;
```
Comment (JacobBarthelmeh) [inline wolfcrypt/src/ecc.c:10516]: This section is a little hard to read, is there no possible way for offset to be larger than the sizeof(key->key_raw) and cause an underflow on the unsigned buf_size?. Please also use cammel case when naming variables, keySize, bufSize.

File: wolfcrypt/src/port/xilinx/xil-aesgcm.c [hunk 3]
```diff
@@ -44,14 +42,272 @@
...
+typedef int (*init_fn)(XSecure_ClientInstance *InstancePtr, XSecure_AesKeySource KeySrc, u32 Size, u64 IvAddr);
+
+enum aad_state {
+    SW_AAD,
+    HW_ENGINE_AAD
+};
+
+static int check_keysize(word32 len)
+{
+    if (len != 16u && len != 32u) {
```
Comment (JacobBarthelmeh) [inline wolfcrypt/src/port/xilinx/xil-aesgcm.c:66]: Lets use AES_128_KEY_SIZE and AES_256_KEY_SIZE enum from wolfssl/wolfcrypt/aes.h instead of the hard set values.

File: wolfcrypt/src/port/xilinx/xil-aesgcm.c [hunk 4]
```diff
@@ -44,14 +42,272 @@
...
+#ifdef WC_XIL_AESGCM_DPA_CM
+    ret = XSecure_AesSetDpaCm(&(aes->xSec.cinst), 1);
+    if (ret != XST_SUCCESS) {
+        WOLFSSL_XIL_ERROR("Failed to enable AES DPA countermeasures", ret);
+        return WC_HW_E;
+    }
+#endif
+    aes->keylen = len;
+    aes->kup = kup;
+    aes->xKeySize = len == 16u ? XSECURE_AES_KEY_SIZE_128 : XSECURE_AES_KEY_SIZE_256;
```
Comment (JacobBarthelmeh) [inline wolfcrypt/src/port/xilinx/xil-aesgcm.c:104]: AES_128_KEY_SIZE instead of 16u

File: wolfcrypt/src/port/xilinx/xil-sha3.c [hunk 5]
```diff
@@ -24,7 +24,7 @@
     #include <config.h>
 #endif
 
-#include <wolfssl/wolfcrypt/settings.h>
+#include "wolfssl/wolfcrypt/settings.h"
```
Comment (JacobBarthelmeh) [inline wolfcrypt/src/port/xilinx/xil-sha3.c:27]: why the change in where the header is searched for?

File: wolfcrypt/src/port/xilinx/xil-versal-trng.c [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+#include "wolfssl/wolfcrypt/error-crypt.h"
+#include "wolfssl/wolfcrypt/logging.h"
+#include "wolfssl/wolfcrypt/port/xilinx/xil-versal-trng.h"
+
+#include <xtrngpsv.h>
+
+#if !defined(HAVE_HASHDRBG)
+/** How often shall the DRBG part be re-seeded */
+#ifndef WC_RESEED_INTERVAL
+#define WC_RESEED_INTERVAL (1000000)
```
Comment (JacobBarthelmeh) [inline wolfcrypt/src/port/xilinx/xil-versal-trng.c:40]: How was this initial number chosen? Is there Xilinx guidelines on it?

File: wolfcrypt/src/port/xilinx/xil-versal-trng.c [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    ret = 0;
+
+out:
+    if (xret) {
+        WOLFSSL_MSG_EX("Xilinx API returned 0x%08x", xret);
+    }
+    return ret;
+}
+
+int wc_versal_trng_reset(void)
```
Comment (JacobBarthelmeh) [inline wolfcrypt/src/port/xilinx/xil-versal-trng.c:161]: Lets add comments above each of these functions, it does not have to be elaborate. At the minimum the success return value expected and a short one line brief.

File: wolfcrypt/src/rsa.c [hunk 8]
```diff
@@ -410,13 +412,20 @@ int wc_InitRsaHw(RsaKey* key)
...
 
     if (key == NULL) {
         return BAD_FUNC_ARG;
     }
 
     mSz = mp_unsigned_bin_size(&(key->n));
+#if defined(WOLFSSL_XILINX_CRYPT_VERSAL)
+    /* Allocate 4 bytes more for the public exponent. */
+    m = (unsigned char*)XMALLOC(WOLFSSL_XSECURE_RSA_KEY_SIZE + 4, key->heap, DYNAMIC_TYPE_KEY);
+    XMEMSET(m, 0, WOLFSSL_XSECURE_RSA_KEY_SIZE + 4);
```
Comment (JacobBarthelmeh) [inline wolfcrypt/src/rsa.c:425]: NULL check on 'm' before memset'ing it.

File: wolfcrypt/src/rsa.c [hunk 9]
```diff
@@ -426,6 +435,9 @@ int wc_InitRsaHw(RsaKey* key)
         XFREE(m, key->heap, DYNAMIC_TYPE_KEY);
         return MP_READ_E;
     }
+#if defined(WOLFSSL_XILINX_CRYPT_VERSAL)
+    XMEMSET(m + mSz, 0, WOLFSSL_XSECURE_RSA_KEY_SIZE + 4 - mSz);
```
Comment (JacobBarthelmeh) [inline wolfcrypt/src/rsa.c:439]: will mSz always be less than WOLFSSL_XSECURE_RSA_KEY_SIZE + 4?

File: wolfcrypt/src/signature.c [hunk 10]
```diff
@@ -198,7 +198,7 @@ int wc_SignatureVerifyHash(
         #if defined(WOLFSSL_SMALL_STACK) && !defined(WOLFSSL_NO_MALLOC)
             byte *plain_data;
         #else
-            byte  plain_data[MAX_ENCODED_SIG_SZ];
+            ALIGN64 byte plain_data[MAX_ENCODED_SIG_SZ];
```
Comment (JacobBarthelmeh) [inline wolfcrypt/src/signature.c:201]: can the need for aligned data be limited to the Xilinx build here?

File: wolfssl/wolfcrypt/aes.h [hunk 11]
```diff
@@ -229,9 +242,16 @@ struct Aes {
     word32  left;            /* unused bytes left from last call */
 #endif
 #ifdef WOLFSSL_XILINX_CRYPT
+#ifdef versal
```
Comment (JacobBarthelmeh) [inline wolfssl/wolfcrypt/aes.h:245]: Lets use the WOLFSSL version of this macro.

File: wolfssl/wolfcrypt/port/xilinx/xil-versal-glue.h [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    WOLFSSL_XSECURE_ECC_NUM
+};
+
+typedef struct Xsecure {
+    ALIGN64 u8 mem[XSECURE_SHARED_MEM_SIZE];
+    XMailbox mb;
+    XSecure_ClientInstance cinst;
+} wc_Xsecure;
+
+int wc_InitXsecure(wc_Xsecure* xsec);
```
Comment (JacobBarthelmeh) [inline wolfssl/wolfcrypt/port/xilinx/xil-versal-glue.h:87]: For functions that we do not want available publicly use WOLFSSL_LOCAL before it's declaration.

Comment (JacobBarthelmeh) [reply wolfssl/wolfcrypt/port/xilinx/xil-versal-glue.h:87]: On the other hand for new, public API, use WOLFSSL_API

File: wolfssl/wolfcrypt/port/xilinx/xil-versal-trng.h [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+
+#ifndef WOLF_XIL_CRYPT_VERSAL_TRNG_H
+#define WOLF_XIL_CRYPT_VERSAL_TRNG_H
+
+#ifdef versal
```
Comment (JacobBarthelmeh) [inline wolfssl/wolfcrypt/port/xilinx/xil-versal-trng.h:26]: Lets include wolfssl/wolfcrypt/types.h and use the WOLFSSL version of the versal macro

File: wolfssl/wolfcrypt/port/xilinx/xil-versal-trng.h [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+#include "xparameters.h"
+#include "xil_types.h"
+#include "xsecure_mailbox.h"
+#include "xsecure_defs.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+int wc_versal_trng_init(byte* nonce, word32 nonceSz);
```
Comment (JacobBarthelmeh) [inline wolfssl/wolfcrypt/port/xilinx/xil-versal-trng.h:35]: Prefer camel case for new functions added in. (exception is if in one of our older files, like ecc, then it would match the pattern existing in the file).

**Review by JacobBarthelmeh — CHANGES_REQUESTED** [review]

File: IDE/XilinxSDK/2022_1/wolfCrypt_example/src/lscript.ld [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+/*                                                                 */
+/* This file is automatically generated by linker script generator.*/
+/*                                                                 */
+/* Version: 2018.3                                                 */
+/*                                                                 */
+/* Copyright (c) 2010-2019 Xilinx, Inc.  All rights reserved.      */
+/*                                                                 */
+/* Description : ARM v8 Linker Script                           */
+/*                                                                 */
+/*******************************************************************/
```
[the remark this answers is not in the document]
Comment (sjaeckel) [reply IDE/XilinxSDK/2022_1/wolfCrypt_example/src/lscript.ld]: Done, c.f. README.md

File: wolfcrypt/src/ecc.c [hunk 16]
```diff
@@ -5214,6 +5264,55 @@ static int _ecc_make_key_ex(WC_RNG* rng, int keysize, ecc_key* key,
...
+        goto error_out;
+    err = xil_mpi_import(key->pubkey.y, key->key_raw + key->dp->size, key->dp->size, key->heap);
+    if (err)
+        goto error_out;
+    err = xil_mpi_import(&key->k, key->priv_key, key->dp->size, key->heap);
+    if (err)
+        goto error_out;
+    err = mp_set(key->pubkey.z, 1);
+    if (err) {
+error_out:
```
[the remark this answers is not in the document]
Comment (sjaeckel) [reply wolfcrypt/src/ecc.c]: Done

File: wolfssl/wolfcrypt/port/xilinx/xil-versal-trng.h (hunk 14 above)
Comment (sjaeckel) [reply wolfssl/wolfcrypt/port/xilinx/xil-versal-trng.h]: Done

File: wolfcrypt/src/signature.c [hunk 17]
```diff
@@ -242,8 +245,7 @@ int wc_SignatureVerifyHash(
             }
     #endif /* WOLFSSL_CRYPTOCELL */
             if (ret != 0) {
-                WOLFSSL_MSG("RSA Signature Verify difference!");
-                ret = SIG_VERIFY_E;
+                WOLFSSL_MSG("RSA Signature Verify failed!");
```
Comment (JacobBarthelmeh) [inline wolfcrypt/src/signature.c:248]: Noting that this no longer explicitly sets the error value but propagates the underlying crypto error...should be fine but worth noting in case it causes troubles.

**Review by JacobBarthelmeh — APPROVED** [review]

**Comment by JacobBarthelmeh** [comment]
Test app and benchmark ran for me on a VMK180 device.

File: wolfssl/wolfcrypt/ecc.h [hunk 18]
```diff
@@ -462,13 +468,17 @@ struct ecc_key {
      * private D concatenated. These are respectively at offset `0`,
      * offset `keysize`, and offset `2 * keysize`.
      */
-    byte key_raw[3 * ECC_MAX_CRYPTO_HW_SIZE];
+    byte keyRaw[3 * ECC_MAX_CRYPTO_HW_SIZE];
```
Comment (dgarske) [inline wolfssl/wolfcrypt/ecc.h:471]: This will break some other ports like `silabs/silabs_ecc.c`. Please keep as `key_raw`.

File: wolfcrypt/test/test.c [hunk 19]
```diff
@@ -10759,9 +10768,9 @@ WOLFSSL_TEST_SUBROUTINE int aesgcm_test(void)
 #endif
 #endif
 
-    byte resultT[sizeof(t1)];
-    byte resultP[sizeof(p) + AES_BLOCK_SIZE];
-    byte resultC[sizeof(p) + AES_BLOCK_SIZE];
+    ALIGN64 byte resultT[sizeof(t1)];
```
Comment (dgarske) [inline wolfcrypt/test/test.c:10771]: Why is `ALIGN64` required?

File: wolfcrypt/src/port/xilinx/xil-versal-trng.c [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+
+   Returns 0 on success
+ */
+int wc_VersalTrngGenerate(byte* output, word32 sz)
+{
+    /* The TRNG always generates exactly 32bytes of output */
+    byte buf[XTRNGPSV_SEC_STRENGTH_BYTES];
+    word32 bytes_generated = 0;
+    do {
+        u32 bytes_left = sz - bytes_generated;
```
Comment (dgarske) [inline wolfcrypt/src/port/xilinx/xil-versal-trng.c:208]: Using a mix of `word32` and `u32` and `int` and `s32` here. Would prefer ours.

File: wolfcrypt/src/port/xilinx/xil-versal-trng.c [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+
+   Returns 0 on success
+ */
+int wc_VersalTrngInit(byte* nonce, word32 nonceSz)
+{
+#if !defined(HAVE_HASHDRBG)
+    /* Use the HRNG mode (with DF) which lets us extract FIPS compliant
+     * random data from the TRNG.
+     */
+    XTrngpsv_UsrCfg user_cfg = {
```
Comment (dgarske) [inline wolfcrypt/src/port/xilinx/xil-versal-trng.c:112]: Can this be `const` too?

File: wolfcrypt/src/ecc.c [hunk 22]
```diff
@@ -5282,6 +5332,52 @@ static int _ecc_make_key_ex(WC_RNG* rng, int keysize, ecc_key* key,
...
+    if (err != XST_SUCCESS) {
+        WOLFSSL_XIL_ERROR("Generate ECC key failed", err);
+        err = WC_HW_E;
+    }
+
+    WOLFSSL_XIL_DCACHE_INVALIDATE_RANGE(XIL_CAST_U64(key->keyRaw), 2 * key->dp->size);
+
+#ifdef WOLFSSL_VALIDATE_ECC_KEYGEN
+    if (err == 0)
+        err = XSecure_EllipticValidateKey(&(key->xSec.cinst), xil_curve_type[key->dp->id], XIL_CAST_U64(key->keyRaw));
```
Comment (dgarske) [inline wolfcrypt/src/ecc.c:5362]: line length 80 max.

File: wolfcrypt/src/ecc.c [hunk 23]
```diff
@@ -5841,7 +5959,76 @@ static int wc_ecc_sign_hash_hw(const byte* in, word32 inlen,
...
+                                           XIL_CAST_U64(key->privKey),
+                                           XIL_CAST_U64(K),
+                                           XIL_CAST_U64(out));
+        if (err) {
+            WOLFSSL_XIL_ERROR("Generate ECC signature failed", err);
+            err = WC_HW_E;
+        }
+
+        WOLFSSL_XIL_DCACHE_INVALIDATE_RANGE(XIL_CAST_U64(out), keysize * 2);
+        fp_reverse(&out[0], keysize);
```
Comment (dgarske) [inline wolfcrypt/src/ecc.c:6018]: Please use the `mp_` functions so it works with all our math library options (not use fast math).

**Review by dgarske — CHANGES_REQUESTED** [review]

**Review by sjaeckel — COMMENTED** [review]
> We will need to squash this commit. Found some whitespace issues:
> 
> ```
> [...]
> ./wolfssl/wolfcrypt/settings.h:2758:»(defined(HAVE_PKCS11) || defined(HAVE_PK_CALLBACKS) || \
> ./wolfssl/wolfcrypt/settings.h:2759:» defined(WOLF_CRYPTO_CB) || defined(WOLFSSL_KCAPI))
> ./wolfssl/wolfcrypt/settings.h:2760:» /* Enables support for using wolfSSL_CTX_use_PrivateKey_Id and
> ./wolfssl/wolfcrypt/settings.h:2761:»  *   wolfSSL_CTX_use_PrivateKey_Label */
> ./wolfssl/wolfcrypt/settings.h:2762:»#define WOLF_PRIVATE_KEY_ID
> ./scripts/bench/bench_functions.sh:18:»[ "$my_path" != "" ] || { echo "\$my_path must not be empty"; return 1; }
> [...]
> ```

1. how is this checked? I couldn't find a way to run that myself. (Also shouldn't this be a CI job!?)
2. those mentioned issues in `./wolfssl/wolfcrypt/settings.h` were already fixed in 0603031362a0f1c819a7bc77258fd7f6a4f9dec9

I've replaced the tabs and fixed the indentation of the changes in this PR. How should I proceed? I can add another commit on top of the existing ones or I've also been able to fixup most of the original commits. What do you prefer?

I'll also wait for an answer to https://github.com/wolfSSL/wolfssl/pull/5166#discussion_r971273450 until I push those latest changes.

File: wolfcrypt/test/test.c (hunk 19 above)
Comment (dgarske) [reply wolfcrypt/test/test.c:10771]: Shouldn't the alignment be handled in the AES GCM call? If not 64-bit aligned then allocate aligned buffer. If you want it in test.c for testing purposes it should not be for ALL cases, just for the Versal build. Make sense?

File: wolfcrypt/src/port/xilinx/xil-versal-trng.c (hunk 20 above)
Comment (dgarske) [reply wolfcrypt/src/port/xilinx/xil-versal-trng.c:208]: My preference is use ours... word32 and int.

File: wolfcrypt/src/ecc.c (hunk 23 above)
Comment (dgarske) [reply wolfcrypt/src/ecc.c:6018]: Can you use `ByteReverseWords`?

Comment (sjaeckel) [reply wolfcrypt/src/ecc.c:6028]: No, because they do something different.

Comment (dgarske) [reply wolfcrypt/src/ecc.c:6028]: I would suggest moving fp_reverse to wolfmath.c and renaming to mp_reverse. If you have trouble let me know and I'll share a patch or push to your PR.

File: wolfcrypt/test/test.c (hunk 19 above)
Comment (sjaeckel) [reply wolfcrypt/test/test.c:10771]: Can I expect `XMALLOC()` to always return an at-least 64bit aligned buffer?

Comment (dgarske) [reply wolfcrypt/test/test.c:10771]: Depends on RTOS, but generally no. You should allocate like this:

See this snippet in aes.c:1809 for AESNI as a reference:
```
        /* check alignment, decrypt doesn't need alignment */
        if ((wc_ptr_t)inBlock % AESNI_ALIGN) {
        #ifndef NO_WOLFSSL_ALLOC_ALIGN
            byte* tmp = (byte*)XMALLOC(AES_BLOCK_SIZE + AESNI_ALIGN, aes->heap,
                                                      DYNAMIC_TYPE_TMP_BUFFER);
            byte* tmp_align;
            if (tmp == NULL)
                return MEMORY_E;

            tmp_align = tmp + (AESNI_ALIGN - ((wc_ptr_t)tmp % AESNI_ALIGN));

            XMEMCPY(tmp_align, inBlock, AES_BLOCK_SIZE);
            AES_ECB_encrypt(tmp_align, tmp_align, AES_BLOCK_SIZE,
                    (byte*)aes->key, aes->rounds);
            XMEMCPY(outBlock, tmp_align, AES_BLOCK_SIZE);
            XFREE(tmp, aes->heap, DYNAMIC_TYPE_TMP_BUFFER);
            return 0;
        #else
            WOLFSSL_MSG("AES-ECB encrypt with bad alignment");
            WOLFSSL_ERROR_VERBOSE(BAD_ALIGN_E);
            return BAD_ALIGN_E;
        #endif
        }
```

File: wolfcrypt/src/port/xilinx/xil-versal-trng.c (hunk 20 above)
Comment (sjaeckel) [reply wolfcrypt/src/port/xilinx/xil-versal-trng.c]: Done

File: wolfcrypt/src/ecc.c (hunk 23 above)
Comment (sjaeckel) [reply wolfcrypt/src/ecc.c]: Done

**Comment by dgarske** [comment]
@sjaeckel 

```
./wolfcrypt/src/include.am:              wolfcrypt/src/port/xilinx/xil-versal-glue.c \
./wolfcrypt/src/include.am:              wolfcrypt/src/port/xilinx/xil-versal-trng.c \
./wolfssl/wolfcrypt/include.am:                         wolfssl/wolfcrypt/port/xilinx/xil-versal-glue.h \
./wolfssl/wolfcrypt/include.am:                         wolfssl/wolfcrypt/port/xilinx/xil-versal-trng.h \
AUTOCONF NEW FILE DETECTOR FAILED!
```

**Comment by dgarske** [comment]
> @sjaeckel
> 
> ```
> ./wolfcrypt/src/include.am:              wolfcrypt/src/port/xilinx/xil-versal-glue.c \
> ./wolfcrypt/src/include.am:              wolfcrypt/src/port/xilinx/xil-versal-trng.c \
> ./wolfssl/wolfcrypt/include.am:                         wolfssl/wolfcrypt/port/xilinx/xil-versal-glue.h \
> ./wolfssl/wolfcrypt/include.am:                         wolfssl/wolfcrypt/port/xilinx/xil-versal-trng.h \
> AUTOCONF NEW FILE DETECTOR FAILED!
> ```

Disregard.. that was a failure from prior commit. It passed now with latest.

**Review by dgarske — APPROVED** [review]
This looks good, but I am unable to test. @JacobBarthelmeh please also review and perhaps @ejohnstown can also review?

File: wolfcrypt/src/wolfmath.c [hunk 24]
```diff
@@ -71,6 +71,23 @@
...
 
+/* reverse an array, used for radix code */
+void mp_reverse (unsigned char *s, int len)
+{
+  int     ix, iy;
+  unsigned char t;
+
+  ix = 0;
+  iy = len - 1;
+  while (ix < iy) {
```
[the remark this answers is not in the document]
Comment (sjaeckel) [reply wolfcrypt/src/wolfmath.c]: fixed

File: wolfcrypt/src/wolfmath.c [hunk 25]
```diff
@@ -71,6 +71,23 @@
 #endif
 
 
+/* reverse an array, used for radix code */
+void mp_reverse (unsigned char *s, int len)
+{
+  int     ix, iy;
+  unsigned char t;
+
+  ix = 0;
```
[the remark this answers is not in the document]
Comment (sjaeckel) [reply wolfcrypt/src/wolfmath.c]: done

**Review by ejohnstown — APPROVED** [review]

**Comment by paulwratt** [comment]
@sjaeckel thanks for getting this done, I was looking for a successful bringup PR

**Comment by sjaeckel** [comment]
> how can I (easily) get a patchset for this (like the link provided in the email notification on the very 1st comment)

If you really need a patch file I guess the easiest would be to download the patch generated by GitHub via https://github.com/wolfSSL/wolfssl/pull/5166.patch

**Comment by paulwratt** [comment]
no, thats the one, its for a reference.

I was just on a thread on the wolf boot and noticed at the bottom of the page it says (like the link you provide) "add .patch or .diff" but it does not say that here .. Thanks

**Review by dgarske — APPROVED** [review]
